external-crates/move/crates/move-vm-runtime/src/loader/package_tables.rs

```diff
@@ -0,0 +1,99 @@
+// Copyright (c) The Move Contributors
+// SPDX-License-Identifier: Apache-2.0
+
+// This module is responsible for the building of the package VTables given a root package storage
+// ID. The VTables are built by loading all the packages that are dependencies of the root package,
+// and once they are loaded creating the VTables for each package, and populating the
+// `loaded_packages` table (keyed by the _runtime_ package ID!) with the VTables for each package
+// in the transitive closure of the root package.
+
+use super::{
+    arena::ArenaPointer,
+    ast::Function,
+    linkage_checker,
+    package_cache::{LoadedPackage, RuntimePackageId, VTableKey},
+    package_loader::PackageLoader,
+    type_cache::TypeCache,
+};
+use move_binary_format::errors::{PartialVMError, PartialVMResult, VMResult};
+use move_core_types::vm_status::StatusCode;
+use move_vm_types::data_store::DataStore;
+use parking_lot::RwLock;
+use std::{collections::HashMap, sync::Arc};
+
+/// The data structure that the VM uses to resolve all packages. Packages are loaded into this at
+/// before the beginning of execution, and based on the static call graph of the package that
+/// contains the root package id.
+///
+/// This is a transient (transaction-scoped) data structure that is created at the beginning of the
+/// transaction, is immutable for the execution of the transaction, and is dropped at the end of
+/// the transaction.
+///
+/// TODO(tzakian): The representation can be optimized to use a more efficient data structure for
+/// vtable/cross-package function resolution but we will keep it simple for now.
+pub struct PackageTables<'a> {
+    loaded_packages: HashMap<RuntimePackageId, Arc<LoadedPackage>>,
```

Comment:
We probably want to replace this with a `CHashMap` or similar on a longer timeline.